src/hooks/useTocIntersection.ts

```diff
@@ -0,0 +1,77 @@
+import { useCallback, useEffect, useRef, useState } from 'react'
+
+interface TocItem {
+  id: string
+  title: string
+  level: number
+  element: HTMLElement
+}
+
+export const useTocIntersection = (tocItems: TocItem[]) => {
+  const [activeId, setActiveId] = useState<string>('')
+  const [highlightedId, setHighlightedId] = useState<string>('')
+  const activeIdRef = useRef<string>('')
+
+  const handleScroll = useCallback(() => {
+    if (tocItems.length === 0) return
+
+    const scrollPosition = window.scrollY
+    let currentSection = ''
+
+    for (let i = 0; i < tocItems.length; i++) {
+      const element = document.getElementById(tocItems[i].id)
+      if (element) {
+        const elementTop = element.offsetTop
+        if (elementTop <= scrollPosition) {
+          currentSection = tocItems[i].id
+        } else {
+          break
+        }
+      }
+    }
+
+    if (currentSection && currentSection !== activeIdRef.current) {
+      activeIdRef.current = currentSection
+      setActiveId(currentSection)
+    }
+  }, [tocItems])
+
+  useEffect(() => {
+    if (tocItems.length === 0) return
+
+    const setInitialState = () => {
+      if (tocItems.length > 0) {
+        activeIdRef.current = tocItems[0].id
+        setActiveId(tocItems[0].id)
+      }
+    }
+
+    setInitialState()
```

Comment:
why create a func and then call it and not just execute the body of the function in the useeffect directly?